### PR TITLE
New module: proxmox_snap

### PR DESCRIPTION
**Pull request by Thulium-Drake**
This commit adds proxmox_snap for snapshot management on PVE, previously worked on in #535 

This is that same module on a new branch (a lot of time passed and I messed up my git checkouts :-) )

##### SUMMARY

proxmox_snap: based on 'proxmox', adapted to manage snapshots for LXC/Qemu machines (OpenVZ might work, but I didn't test it as it is no longer supported in PVE)

Status: fully functional, lacks tests

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

proxmox_snap

### CODE REVIEW

**Comment by Thulium-Drake** [comment]
cc @felixfontein

**Comment by Thulium-Drake** [comment]
I copied the only test in test_terraform.py and renamed it to work for this module. But I'm a bit stuck at _what_ I should test, I tried looking at other tests, but it still is a kind (the kind I don't get just yet) of magic to me

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: proxmox_snap
+short_description: snapshot management of instances in Proxmox VE cluster
+description:
+  - allows you to create/delete snapshots from instances in Proxmox VE cluster
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_snap.py:13]:
```suggestion
  - Allows you to create/delete snapshots from instances in Proxmox VE cluster.
version_added: 1.3.0
```

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+options:
+  api_host:
+    description:
+      - the host of the Proxmox VE cluster
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_snap.py:17]:
```suggestion
      - The host of the Proxmox VE cluster.
```
Same for the other descriptions. Every paragraph (list item) should be a (or multiple) correct sentence(s).

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    description:
+      - Snapshot includes RAM
+    default: 'no'
+    type: bool
+  description:
+    description:
+      - Specify the description for the snapshot. Only used on the configuration web interface.
+      - This is saved as a comment inside the configuration file.
+    type: str
+    version_added: '0.2.0'
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_snap.py:65]:
```suggestion
```
Same for the other below.

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+    type: str
+  force:
+    description:
+      - For removal from config file, even if removing disk snapshot fails.
+    default: 'no'
+    type: bool
+  vmstate:
+    description:
+      - Snapshot includes RAM
+    default: 'no'
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_snap.py:58]:
```suggestion
    default: no
```
Same above.

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+notes:
+  - Requires proxmoxer and requests modules on host. This modules can be installed with pip.
+requirements: [ "proxmoxer", "python >= 2.7", "requests" ]
+author: Jeffrey van Pelt (@Thulium-Drake)
+'''
+
+EXAMPLES = r'''
+- name: Create new container snapshot
+  proxmox_snap:
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_snap.py:86]: Always use FQCN in examples:
```suggestion
  community.general.proxmox_snap:
```

File: tests/unit/plugins/modules/cloud/misc/test_proxmox_snap.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# Copyright: (c) 2019, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import json
+
+import pytest
+
+proxmox = pytest.importorskip('proxmoxer')
```
Comment (felixfontein) [inline tests/unit/plugins/modules/cloud/misc/test_proxmox_snap.py:10]: That won't help much for CI, since that package isn't around. It would be better if you could mock the ProxmoxAPI object.

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 5 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py]: fixed!

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 4 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py]: fixed

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 3 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py]: fixed!

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 2 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py]: fixed!

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 1 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py]: fixed!

**Comment by Thulium-Drake** [comment]
I'm having a really hard time mocking up the ProxmoxAPI object, I think I know what to do when I have that, I 'just' need to fake a few functions of it to supply API responses from a PVE server.

Anyone have any pointers on how to fake an object like:

```
from proxmoxer import ProxmoxAPI
proxmox = ProxmoxAPI(args)
```

I need to bypass the importer to actually believe there is a ```proxmoxer``` module, and I need that module to contain that ProxmoxAPI object

**Comment by felixfontein** [comment]
You could do it similar to here:
- Plugin code, add factory function: https://github.com/ansible-collections/community.general/blob/main/plugins/lookup/dsv.py#L106-L108
- Test code: modify "import worked" variable (https://github.com/ansible-collections/community.general/blob/main/tests/unit/plugins/lookup/test_dsv.py#L28) and patch the factory function: https://github.com/ansible-collections/community.general/blob/main/tests/unit/plugins/lookup/test_dsv.py#L31-L34

**Comment by felixfontein** [comment]
(Of course you can also create a free function for that, instead of a class/static method.)

**Comment by felixfontein** [comment]
In your last force push, you undid a lot of good changes to the module docs itself.

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+      community.general.proxmox_snap:
+        api_user: root@pam
+        api_password: 1q2w3e
+        api_host: node1
+        vmid: "{{ proxmox_vmid }}"
+        state: present
+        snapname: test
+        vmstate: true
+      delegate_to: localhost
+    - name: remove snapshots using the inventory hostname
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:115]:
```suggestion
      delegate_to: localhost
  
    - name: Remove snapshots using the inventory hostname
```
and is `delegated_to: localhost` really needed here?

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+    api_host: node1
+    vmid: 100
+    state: absent
+    snapname: pre-updates
+
+- name: Make snapshots of multiple machines
+  hosts: all
+  gather_facts: false
+  tasks:
+    - name: make snapshots using the vmid
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:105]:
```suggestion
    - name: Make snapshots using the vmid
```
according to Ansible dev guide

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+            state=dict(default='present', choices=['present', 'absent']),
+            description=dict(type='str'),
+            snapname=dict(type='str', default='ansible_snap'),
+            force=dict(type='bool', default='no'),
+            vmstate=dict(type='bool', default='no'),
+        )
+    )
+
+    if not HAS_PROXMOXER:
+        module.fail_json(msg='proxmoxer required for this module')
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:209]:
```suggestion
        module.fail_json(msg='proxmoxer required for this module')
```
use `missed_required_lib` here, grep through other stuff

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+        try:
+            api_password = os.environ['PROXMOX_PASSWORD']
+        except KeyError as e:
+            module.fail_json(msg='You should set api_password param or use PROXMOX_PASSWORD environment variable')
+
+    try:
+        proxmox = setup_api(api_host, api_user, api_password, validate_certs)
+
+    except Exception as e:
+        module.fail_json(msg='authorization on proxmox cluster failed with exception: %s' % e)
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:235]:
```suggestion
        module.fail_json(msg='authorization on proxmox cluster failed with exception: %s' % to_native(e))
```
this prevents python-version-related errors

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+    snapname = module.params['snapname']
+    timeout = module.params['timeout']
+    force = module.params['force']
+    vmstate = module.params['vmstate']
+
+    # If password not set get it from PROXMOX_PASSWORD env
+    if not api_password:
+        try:
+            api_password = os.environ['PROXMOX_PASSWORD']
+        except KeyError as e:
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:228]:
```suggestion
        except KeyError:
```
or better put `to_native(e)` to the error message

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+                module.fail_json(msg='VM with vmid = %s not exists in cluster' % vmid)
+
+            for i in getattr(proxmox.nodes(vm[0]['node']), VZ_TYPE)(vmid).snapshot.get():
+                if i['name'] == snapname:
+                    module.exit_json(changed=False, msg="Snapshot %s is already present" % snapname)
+
+            if snapshot_create(module, proxmox, vm, vmid, timeout, snapname, description, vmstate):
+                module.exit_json(changed=True, msg="Snapshot %s created" % snapname)
+        except Exception as e:
+            module.fail_json(msg="Creating snapshot %s of VM %s failed with exception: %s" % (snapname, vmid, e))
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:264]:
```suggestion
            module.fail_json(msg="Creating snapshot %s of VM %s failed with exception: %s" % (snapname, vmid, to_native(e)))
```

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+                    snap_exist = True
+                    continue
+
+            if not snap_exist:
+                module.exit_json(changed=False, msg="Snapshot %s does not exist" % snapname)
+            else:
+                if snapshot_remove(module, proxmox, vm, vmid, timeout, snapname, force):
+                    module.exit_json(changed=True, msg="Snapshot %s removed" % snapname)
+        except Exception as e:
+            module.fail_json(msg="Removing snapshot %s of VM %s failed with exception: %s" % (snapname, vmid, e))
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:285]:
```suggestion
            module.fail_json(msg="Removing snapshot %s of VM %s failed with exception: %s" % (snapname, vmid, to_native(e)))
```

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 14]
```diff
@@ -0,0 +1,289 @@
+#!/usr/bin/python
+# Copyright: Ansible Project
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:2]:
```suggestion
# Copyright: (c) 2020, Jeffrey van Pelt (@Thulium-Drake) <yourmail@here>
```
just a suggestion

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: community.general.proxmox_snap
+short_description: Snapshot management of instances in Proxmox VE cluster.
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:11]:
```suggestion
short_description: Snapshot management of instances in Proxmox VE cluster
```

**Review by Andersson007 — COMMENTED** [review]
@Thulium-Drake hi, thanks for the PR!
```
/root/test.py:129:1: F401 'traceback' imported but unused
```
and what about `check_mode` support? If it's possible to implement, would be great to have this. Also add info about `check_mode` to `note:`, e.g.
`
Supports C(check_mode).
`
or `Doesn't support C(check_mode).` resp

Thanks!

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 16]
```diff
@@ -0,0 +1,289 @@
+#!/usr/bin/python
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:1]:
```suggestion
#!/usr/bin/python
# -*- coding: utf-8 -*-

```

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 7 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py:115]: Not really required, but as this module has an external dependency on the ```proxmoxer``` library, it makes sense to install that only on the Ansible controller. It should not be run on the 'target' host itself.

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 8 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py:105]: fixed

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 10 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py:235]: gotcha! fixed! :-)

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 9 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py]: Done!

**Comment by Thulium-Drake** [comment]
@Andersson007 Hi Andrew, thanks a million for reviewing my code! Sorry it took me so long to get back to it. I added, I think, a decent check mode, it will run through all the logic and then chicken out at the end of each code path. Is that what you meant? :-)

Thanks!

**Comment by Thulium-Drake** [comment]
ready_for_review

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 17]
```diff
@@ -0,0 +1,310 @@
...
+      community.general.proxmox_snap:
+        api_user: root@pam
+        api_password: 1q2w3e
+        api_host: node1
+        hostname: "{{ inventory_hostname }}"
+        state: absent
+        snapname: test
+        force: true
+      delegate_to: localhost
+'''
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:128]:
```suggestion
'''

RETURN = r'''#'''
```

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 18]
```diff
@@ -0,0 +1,310 @@
...
+        proxmox = setup_api(api_host, api_user, api_password, validate_certs)
+
+    except Exception as e:
+        module.fail_json(msg='authorization on proxmox cluster failed with exception: %s' % to_native(e))
+
+    # If hostname is set get the VM id from ProxmoxAPI
+    if not vmid and hostname:
+        hosts = get_vmid(proxmox, hostname)
+        if len(hosts) == 0:
+            module.fail_json(msg="Vmid could not be fetched => Hostname doesn't exist (action: %s)" % state)
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:254]:
```suggestion
            module.fail_json(msg="Vmid could not be fetched => Hostname does not exist (action: %s)" % state)
```

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 19]
```diff
@@ -0,0 +1,310 @@
...
+    - name: Make snapshots using the vmid
+      community.general.proxmox_snap:
+        api_user: root@pam
+        api_password: 1q2w3e
+        api_host: node1
+        vmid: "{{ proxmox_vmid }}"
+        state: present
+        snapname: test
+        vmstate: true
+      delegate_to: localhost
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:116]:
```suggestion
```
is it really necessary here as well as `gather_facts` and `hosts: all`?
I'd even remove the `- name: Make snapshots of multiple machines` and put the tasks on the top level because the first one is about `Make snapshots using the vmid` and the second sounds `Remove snapshots using the inventory hostname` which seem to be unrelated, isn't it?

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 20]
```diff
@@ -0,0 +1,310 @@
...
+    default: 30
+    type: int
+  snapname:
+    description:
+      - Name of the snapshot that has to be created.
+    default: 'ansible_snap'
+    type: str
+
+notes:
+  - Requires proxmoxer and requests modules on host. This modules can be installed with pip.
```
Comment (Andersson007) [inline plugins/modules/cloud/misc/proxmox_snap.py:79]:
```suggestion
  - Requires proxmoxer and requests modules on host. This modules can be installed with pip.
  - Supports C(check_mode).
```

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 17 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py:128]: fixed!

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 20 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py:79]: fixed!

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 19 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py:116]: True, both tasks are redundant with regards to the first 2 tasks. It was more meant as an example play on how to use this module (I have a playbook with code similar to this to make snapshots before I install patches)

As mentioned before, just like the VMWare modules, it's not really usefull to run it on all the targeted machines. But only on the Ansible Controller, which should then have access to the Proxmox host running the containers.

And one can provide the Proxmox VMID if the Proxmox inventory is used, or, one can use the VM name in Proxmox and it will figure out the VMID based on that.

I think the best decision is, cut it or keep it. I'm leaning towards cutting, as it may already have caused some confusion. What's your take?

File: plugins/modules/cloud/misc/proxmox_snap.py [hunk 21]
```diff
@@ -0,0 +1,313 @@
...
+# Copyright: (c) 2020, Jeffrey van Pelt (@Thulium-Drake) <jeff@vanpelt.one>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: proxmox_snap
+short_description: Snapshot management of instances in Proxmox VE cluster
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_snap.py:13]:
```suggestion
short_description: Snapshot management of instances in Proxmox VE cluster
version_added: 2.0.0
```

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 19 above)
Comment (felixfontein) [reply plugins/modules/cloud/misc/proxmox_snap.py:117]: Spontaneously, cutting sounds good :)

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 21 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py:13]: Done!

File: plugins/modules/cloud/misc/proxmox_snap.py (hunk 19 above)
Comment (Thulium-Drake) [reply plugins/modules/cloud/misc/proxmox_snap.py:117]: Okido, done! :-)

**Review by felixfontein — APPROVED** [review]

**Comment by Aversiste** [comment]
Hello @Thulium-Drake,

The compatibility with both LXC and Qemu/KVM instances is mentioned in this PR but not in the module's description, and I think it could be a good addition.
What do you think ?

**Comment by Thulium-Drake** [comment]
I agree, added it :-)

**Review by Andersson007 — APPROVED** [review]

**Comment by Thulium-Drake** [comment]
Does anyone still need to add a ```shipit``` before it can be merged?

**Comment by Andersson007** [comment]
@Aversiste how is it for you now?

**Comment by Aversiste** [comment]
Its good for me :)

**Review by Aversiste — APPROVED** [review]

**Comment by felixfontein** [comment]
@Thulium-Drake thanks for this new module!
@Andersson007 @Aversiste thanks for reviewing this!
